dotnet/src/SemanticKernel/Kernel.cs

```diff
@@ -194,6 +201,20 @@ public async Task<SKContext> RunAsync(ContextVariables variables, CancellationTo
             try
             {
                 cancellationToken.ThrowIfCancellationRequested();
+                var functionDetails = f.Describe();
+                string? renderedPrompt = null;
+
+                if (f is SemanticFunction semanticFunction)
+                {
+                    renderedPrompt = await semanticFunction._promptTemplate.RenderAsync(context, cancellationToken).ConfigureAwait(false);
+                }
```

Comment:
renderedPrompt is only used as part of the FunctionInvoking handler?  Does that means we'll end up rendering it twice, once here for the event and another time inside of the actual invocation?  Three issues:
1. Duplicated effort.
2. Couldn't that mean we actually get a different prompt here than what's actually submitted, if there's any non-determinism involved? Or is it always 100% deterministic?
3. We're doing the work here even if there's no handler registered for FunctionInvoking.

Reply:
@stephentoub, that's a good point.
1. We are planning (very soon) to move PromptTemplateEngine logic away from the function execution and actually pass the generated prompt as an input to the function itself in the meanwhile, yes, this is a duplicated effort.

2. It can be potentially non deterministic if the template actually calls other semantic functions (non deterministic by nature) or Native skills like random number generators. That said, those scenarios aren't common.

3. That's new for me, I thought the `FunctionInvoking?.Invoke` would do the null check before trying to call it and not be expensive.

Reply:
As a way to overcome numbers 1 and 2, I added an `internal GetAndForgetRenderedPrompt` method to SKContext which will preserve the GeneratedPrompt in the InvokeAsync, so this way, when using both `IKernel.RunAsync` and `ISKFunction.InvokeAsync` no duplicated effort will happen as well as any problems that could arise from it.